### PR TITLE
 rpc: implement gorilla websocket 

### PR DESCRIPTION
**Pull request by jiseongnoh**
## Proposed changes

Changes websocket library of Klaytn to Gorilla Websocket
Enables HTTP basic authentication
Fixes issue #1111

## Types of changes

Please put an x in the boxes related to your change.

- [x] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: networks/rpc/websocket.go [hunk 1]
```diff
@@ -103,6 +114,72 @@ var upgrader = fastws.Upgrader{
 	WriteBufferSize: 1024,
 }
 
+// GSWebsocketHandler returns a handler that serves JSON-RPC to WebSocket connections.
+//
+// allowedOrigins should be a comma-separated list of allowed origin URLs.
+// To allow connections with any origin, pass "*".
+func (s *Server) GSWebsocketHandler(allowedOrigins []string) http.Handler {
+	var upgrader = gorillaws.Upgrader{
+		ReadBufferSize:  wsReadBuffer,
```
Comment (yoomee1313) [inline networks/rpc/websocket.go:123]: Can you change the fastws.Upgrader too?

File: networks/rpc/client_test.go [hunk 2]
```diff
@@ -492,7 +492,8 @@ func TestClientReconnect(t *testing.T) {
 		if err != nil {
 			t.Fatal(err)
 		}
-		go http.Serve(l, srv.WebsocketHandler([]string{"*"}))
+		//go http.Serve(l, srv.WebsocketHandler([]string{"*"}))
```
Comment (yoomee1313) [inline networks/rpc/client_test.go:495]: Please remove the commented line.

File: networks/rpc/client.go [hunk 3]
```diff
@@ -733,6 +903,42 @@ func (c *Client) read(conn net.Conn) error {
...
+
+	readMessage := func() (rs []*jsonrpcMessage, err error) {
+		_, reader, err := conn.NextReader()
+		if err != nil {
+			return nil, err
+		}
+		var (
+			buf json.RawMessage
+			dec = json.NewDecoder(reader)
+		)
```
Comment (yoomee1313) [inline networks/rpc/client.go:918]: Can I ask why you move those line to here?

File: networks/rpc/client.go [hunk 4]
```diff
@@ -733,6 +903,42 @@ func (c *Client) read(conn net.Conn) error {
 	}
 }
 
+// readGorillaWS is called on a dedicated goroutine.
+
```
Comment (yoomee1313) [inline networks/rpc/client.go:907]:
```suggestion
```

File: networks/rpc/client.go [hunk 5]
```diff
@@ -638,6 +710,104 @@ func (c *Client) dispatch(conn net.Conn) {
 	}
 }
 
+// dispatchGorillaWS is the main loop of the client that uses Gorilla websocket.
+// It sends readGorillaWS messages to waiting calls to Call and BatchCall
+// and subscription notifications to registered subscriptions.
+func (c *Client) dispatchGorillaWS(conn *gorillaws.Conn) {
```
Comment (yoomee1313) [inline networks/rpc/client.go:716]: What about combining `dispatch` and `dispatchGorillaWS` functions together? I think quite a few lines are dulicated.

File: networks/rpc/client.go (hunk 3 above)
Comment (jiseongnoh) [reply networks/rpc/client.go:918]: Variable dec needs a reader. The reader needs the result of conn.NextReader. Thus, var dec must be inside of readMessage() method.

Comment (jiseongnoh) [reply networks/rpc/client.go:918]: Variable dec needs a reader. The reader needs the result of conn.NextReader. Thus, var dec must be inside of readMessage() method.

File: networks/rpc/websocket.go (hunk 1 above)
Comment (jiseongnoh) [reply networks/rpc/websocket.go:123]: Do you mean this?
```diff
diff --git a/networks/rpc/websocket.go b/networks/rpc/websocket.go
index c8bde553..39583df8 100644
--- a/networks/rpc/websocket.go
+++ b/networks/rpc/websocket.go
@@ -109,11 +109,6 @@ func (srv *Server) WebsocketHandler(allowedOrigins []string) http.Handler {
        }
 }

-var upgrader = fastws.Upgrader{
-       ReadBufferSize:  1024,
-       WriteBufferSize: 1024,
-}
-
 // GSWebsocketHandler returns a handler that serves JSON-RPC to WebSocket connections.
 //
 // allowedOrigins should be a comma-separated list of allowed origin URLs.
@@ -182,6 +177,10 @@ func (s *Server) GSWebsocketHandler(allowedOrigins []string) http.Handler {

 func (srv *Server) FastWebsocketHandler(ctx *fasthttp.RequestCtx) {

+       var upgrader = fastws.Upgrader{
+               ReadBufferSize:  1024,
+               WriteBufferSize: 1024,
+       }
        // TODO-Klaytn handle websocket protocol
        protocol := ctx.Request.Header.Peek("Sec-WebSocket-Protocol")
        if protocol != nil {
@@ -256,6 +255,10 @@ func NewGSWSServer(allowedOrigins []string, srv *Server) *http.Server {
 }

 func NewFastWSServer(allowedOrigins []string, srv *Server) *fasthttp.Server {
+       var upgrader = fastws.Upgrader{
+               ReadBufferSize:  1024,
+               WriteBufferSize: 1024,
+       }
        upgrader.CheckOrigin = wsFastHandshakeValidator(allowedOrigins)

        // TODO-Klaytn concurreny default (256 * 1024), goroutine limit (8192)
```

**Pull request closed without merging**